src/actions/write/initiateFoundationAction.ts

```diff
@@ -0,0 +1,165 @@
+import {
+  ALLOWED_ACTIVE_TIERS,
+  DEFAULT_INVALID_ID_TIER_MESSAGE,
+  DEFAULT_INVALID_TIER_MESSAGE,
+  FOUNDATION_ACTION_ACTIVE_STATUS,
+  FOUNDATION_ACTION_PASSED_STATUS,
+  MAX_FOUNDATION_ACTION_PERIOD,
+  MAX_NAME_LENGTH,
+  MAX_NOTE_LENGTH,
+} from '../../constants';
+import {
+  ActiveTier,
+  ContractResult,
+  FoundationAction,
+  IOState,
+  PstAction,
+  ServiceTier,
+} from '../../types';
+import { isValidArweaveBase64URL } from '../../utilities';
+
+declare const ContractError;
+declare const SmartWeave: any;
+
+// Proposes a foundation action
+export const initiateFoundationAction = async (
+  state: IOState,
+  { caller, input: { type, note, value } }: PstAction,
+): Promise<ContractResult> => {
+  const foundation = state.foundation;
+
+  // The caller must be in the foundation, or else this action cannot be initiated
+  if (!foundation.addresses.includes(caller)) {
+    throw new ContractError(
+      `${caller} Caller needs to be in the foundation wallet list.`,
+    );
+  }
+
+  if (typeof note !== 'string' || note.length > MAX_NOTE_LENGTH) {
+    throw new ContractError('Note format not recognized.');
+  }
+
+  if (type === 'addAddress' && typeof value === 'string') {
+    if (!isValidArweaveBase64URL(value)) {
+      throw new ContractError(
+        'The target of this action is an invalid Arweave address?"',
+      );
+    }
```

Comment:
we can likely add some helpers to validate these too

```
export type WalletAddress = string;
export function isValidAddress(value: any): value is WalletAddress {
    return typeof value === 'string' && isValidArweaveBase64URL(value)
}

export type ValidNumberInput = number;
export function isValidNumber(value: any): value is ValidNumberInput {
    return typeof value === 'number' && Number.isInteger(MinmimumSignaturesInput)  && value > 0
}

export FeesInput = { [nameLength: number]: number }
export function isValidFees(value: any): value is FeesInput {
       return typeof value === "object" &&  Object.entries(value).every((key, value) => inputValidNumber(+key) && inputIsValidNumber(+value) && Object.entries(keys).length < MAX_NAME_LENGTH;
}

...
```

Reply:
i'll like propose this in a separate PR - there are multiple functions that can check for it